primary/src/block_synchronizer/handler.rs

```diff
@@ -0,0 +1,248 @@
+        }
+    }
+}
+
+#[automock]
+#[async_trait]
+pub trait Handler<PublicKey: VerifyingKey> {
+    /// It retrieves the requested blocks via the block_synchronizer making
+    /// sure though that they are fully validated. The certificates will only
+    /// be returned when they have properly processed via the core module
+    /// and made sure all the requirements have been fulfilled.
+    async fn get_and_synchronize_block_headers(
+        &self,
+        block_ids: Vec<CertificateDigest>,
+    ) -> Vec<Result<Certificate<PublicKey>, Error>>;
+
+    /// It retrieves the requested blocks via the block_synchronizer, but it
+    /// doesn't synchronize the fetched headers, meaning that no processing
+    /// will take place (causal completion etc).
+    async fn get_block_headers(
+        &self,
+        block_ids: Vec<CertificateDigest>,
+    ) -> Vec<BlockSynchronizeResult<BlockHeader<PublicKey>>>;
+}
+
+/// A helper struct to allow us access the block_synchronizer in a synchronous
+/// way. It also offers methods to both fetch the certificates and way to
+/// process them and causally complete their history.
+pub struct BlockSynchronizerHandler<PublicKey: VerifyingKey> {
+    /// Channel to send commands to the block_synchronizer.
+    tx_block_synchronizer: Sender<Command<PublicKey>>,
+
+    /// Channel to send the fetched certificates to Core for
+    /// further processing, validation and possibly causal
+    /// completion.
+    tx_core: Sender<PrimaryMessage<PublicKey>>,
+
+    /// The store that holds the certificates.
+    certificate_store: Store<CertificateDigest, Certificate<PublicKey>>,
+
+    /// The timeout while waiting for a certificate to become available
+    /// after submitting for processing to core.
+    certificate_deliver_timeout: Duration,
+}
+
+impl<PublicKey: VerifyingKey> BlockSynchronizerHandler<PublicKey> {
+    pub fn new(
+        tx_block_synchronizer: Sender<Command<PublicKey>>,
+        tx_core: Sender<PrimaryMessage<PublicKey>>,
+        certificate_store: Store<CertificateDigest, Certificate<PublicKey>>,
+        certificate_deliver_timeout: Duration,
+    ) -> Self {
+        Self {
+            tx_block_synchronizer,
+            tx_core,
+            certificate_store,
+            certificate_deliver_timeout,
+        }
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    async fn wait_all(
+        &self,
+        certificates: Vec<Certificate<PublicKey>>,
+    ) -> Vec<Result<Certificate<PublicKey>, Error>> {
+        let futures: Vec<_> = certificates
+            .into_iter()
+            .map(|c| self.wait(c.digest()))
+            .collect();
+
+        join_all(futures).await
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    async fn wait(&self, block_id: CertificateDigest) -> Result<Certificate<PublicKey>, Error> {
+        if let Ok(result) = timeout(
+            self.certificate_deliver_timeout,
+            self.certificate_store.notify_read(block_id),
```

Comment:
I am using here the `certificate_store` to wait and get the "signal" the the certificate has been processed. However, I would like to replace this eventually by the [certificate_waiter](https://github.com/MystenLabs/narwhal/blob/main/primary/src/certificate_waiter.rs) which deals already with it. Unfortunately the [certificate_waiter](https://github.com/MystenLabs/narwhal/blob/main/primary/src/certificate_waiter.rs) doesn't offer a blocking API and also doesn't allow us to channel the outcome to a provide channel (only outputs to core). Until we have this available, I am using the `certificate_store`

Reply:
I think this is worth opening an issue.

Reply:
Opened here https://github.com/MystenLabs/narwhal/issues/232